fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java

```diff
@@ -3615,10 +3615,11 @@ public ChangedTransactionDetail adjustExistingTransaction(final LoanTransaction
         validateActivityNotBeforeClientOrGroupTransferDate(LoanEvent.LOAN_REPAYMENT_OR_WAIVER,
                 transactionForAdjustment.getTransactionDate());
 
-        if (transactionForAdjustment.isNotRepayment() && transactionForAdjustment.isNotWaiver()) {
-            final String errorMessage = "Only transactions of type repayment or waiver can be adjusted.";
-            throw new InvalidLoanTransactionTypeException("transaction", "adjustment.is.only.allowed.to.repayment.or.waiver.transaction",
-                    errorMessage);
+        if (transactionForAdjustment.isNotRepayment() && transactionForAdjustment.isNotWaiver()
```

Comment:
This is rather a comment for the changes on the whole file and not specific to this line.

John, do you think it would be possible to extract all these changes to a separate class instead of putting everything into the Loan class?

I brought this up on the mailing list as well that unless we have a really really good reason, we must start extracting new functionality outside of this class to decrease the maintenance cost.